keps/0023-enable-disable-features.md

```diff
@@ -0,0 +1,131 @@
+And when talking about enabling/disabling features, we are eventually talking about the `Kind` of the Task. Currently, we have `Apply`, `Delete` and `Pipe` tasks kind.
+Enabling a feature is running a task of `Apply` kind and disabling is running a task of `Delete`
+
+We propose introducing a new Task kind named `Feature`. The Feature kind handles Apply or Delete resources based on the value of a parameter in the Operator Instance.
+
+A `Feature` task will look like:
+
+```
+tasks:
+  - name: mirror-maker-configuration
+    kind: Feature
+    spec:
+      parameter: MIRROR_MAKER_ENABLED
+      resources:
+        - mirror-maker-cm.yaml
+  - name: mirror-maker
+    kind: Feature
+    spec:
+      parameter: MIRROR_MAKER_ENABLED
+      resources:
+        - mirror-maker.yaml
+```
+
+
+the parameter used in task would be defined in the parameters for the Operator, and must evaluate to a boolean value.
+```
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: MIRROR_MAKER_ENABLED
+    default: "false"
+    trigger: mirror-maker-plan
+```
+
+And the plan triggered by the parameter will be the one using `Feature` tasks in their steps.
+
+```
+plans:
+  mirror-maker-plan:
+    strategy: serial
+    phases:
+      - name: deploy-mirror-maker
+        strategy: serial
+        steps:
+          - name: configuration
+            tasks:
+              - mirror-maker-configuration
+          - name: deploy
+            tasks:
+              - mirror-maker
+```
+
+In the tasks is where we can use the `Feature` kind. Which will apply the resources rendered by templates `mirror-maker.yaml` and `mirror-maker-cm.yaml` in case the value of parameter `MIRROR_MAKER_ENABLED` is true. 
+In case, the value is `false` it will delete the resources rendered by `mirror-maker.yaml` and `mirror-maker-cm.yaml`. 
+
+In case the parameter value is **true**:
+
+If there are objects to created the Task will be marked successful, once those are created.
+If there are no objects to be created the Task will be marked as successful. 
+
+In case the parameter value is **false**:
+
+If there are objects to be deleted the Task will be marked as successful, once those are marked for Deletion. 
+If there are no objects to be deleted the Task will be marked as successful. 
+
+
+### Limitations
+
+However the `Feature` task will have limitations like:
+
+- KUDO Operator must define the parameter which will be evaluated as a boolean.
+- The defined parameter must evaluate to a boolean value
+- Multiple parameters can trigger the same plan, but only one parameter will be evaluated
+
+For the last use case, e.g., we can have a parameter `MIRROR_MAKER_MEM` which triggers the `mirror-maker-plan` but
+the parameter which will decide in `Feature Task` if the kubernetes objects are created/updated or deleted will be `MIRROR_MAKER_ENABLED`
+
+## Alternatives
+
+There are two alternatives to this approach
```

Comment:
Here is another alternative:
The current suggestion encapsulates the `if-else` logic in a new task type. However, both branches (_true_ and _false_) are already covered by the `Apply` and `Delete` tasks so we might as well put the `if-else` logic directly into the plan like:

```
plans:
mirror-maker-plan:
  strategy: serial
  phases:
    - name: mirror-maker-phase
      strategy: serial
      steps:
        - name: apply-mirror-maker
          tasks:
            - apply-mirror-maker-configuration
            - apply-mirror-maker
          when: (MIRROR_MAKER_ENABLED == "true")
        - name: delete-mirror-maker
          tasks:
            - delete-mirror-maker-configuration
            - delete-mirror-maker
          when: (MIRROR_MAKER_ENABLED == "false")
```
Here, e.g. `apply-mirror-maker` is a simple `Apply` and `delete-mirror-maker` a `Delete` kind of task (omited here for brevity). 

**Pros:**
 - we don't need a new task type
 - more conditions flexibility: it's easy to extend the `when` to whatever we need it to be

**Cons:**
 - YAMLy programming language 🤮
 - hard to debug

P.S. I'm not advocating for it. I think it is awful  😂 But It is semantically the same and makes it clear that we're not in the "declarative world" anymore.

Reply:
Good point for an alternative. And it sounds tempting, because it's kind of powerful, but as you said, hard to debug, and if we want programming, it shouldn't be in yaml.

Reply:
added the alternative!